gpio/zynq7000-gpio/gpio.c

```diff
@@ -0,0 +1,296 @@
+/* GPIO pins */
+const int gpioPins[GPIO_BANKS][GPIO_PINS] = {
+	{
+		GPIO0_0, GPIO0_1, GPIO0_2, GPIO0_3,
+		GPIO0_4, GPIO0_5, GPIO0_6, GPIO0_7,
+		GPIO0_8, GPIO0_9, GPIO0_10, GPIO0_11,
+		GPIO0_12, GPIO0_13, GPIO0_14, GPIO0_15,
+		GPIO0_16, GPIO0_17, GPIO0_18, GPIO0_19,
+		GPIO0_20, GPIO0_21, GPIO0_22, GPIO0_23,
+		GPIO0_24, GPIO0_25, GPIO0_26, GPIO0_27,
+		GPIO0_28, GPIO0_29, GPIO0_30, GPIO0_31
+	},
+	{
+		GPIO1_0, GPIO1_1, GPIO1_2, GPIO1_3,
+		GPIO1_4, GPIO1_5, GPIO1_6, GPIO1_7,
+		GPIO1_8, GPIO1_9, GPIO1_10, GPIO1_11,
+		GPIO1_12, GPIO1_13, GPIO1_14, GPIO1_15,
+		GPIO1_16, GPIO1_17, GPIO1_18, GPIO1_19,
+		GPIO1_20, GPIO1_21, -1, -1,
+		-1, -1, -1, -1,
+		-1, -1, -1, -1,
+	},
+};
+
+
+static struct {
+	volatile uint32_t *base; /* GPIO registers base address */
+} gpio_common;
+
+
+static int gpio_checkPin(unsigned int bank, unsigned int pin)
+{
+	if ((bank >= GPIO_BANKS) || (pin >= GPIO_PINS)) {
+		return -EINVAL;
+	}
+
+	if (gpioPins[bank][pin] < 0) {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_readPort(unsigned int bank, uint32_t *val)
+{
+	if (bank == 0) {
+		*val = *(gpio_common.base + rodata0);
+	}
+	else if (bank == 1) {
+		*val = *(gpio_common.base + rodata1);
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_writePort(unsigned int bank, unsigned char word, uint32_t maskval)
+{
+	if (bank == 0) {
+		/* Write MSW */
+		if (word) {
+			*(gpio_common.base + dmsw0) = maskval;
+		}
+		/* Write LSW */
+		else {
+			*(gpio_common.base + dlsw0) = maskval;
+		}
+	}
+	else if (bank == 1) {
+		/* Write MSW */
+		if (word) {
+			*(gpio_common.base + dmsw1) = maskval;
+		}
+		/* Write LSW */
+		else {
+			*(gpio_common.base + dlsw1) = maskval;
+		}
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_readDir(unsigned int bank, uint32_t *dir)
+{
+	if (bank == 0) {
+		*dir = *(gpio_common.base + dirm0);
+	}
+	else if (bank == 1) {
+		*dir = *(gpio_common.base + dirm1);
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_writeDir(unsigned int bank, uint32_t dir, uint32_t mask)
+{
+	uint32_t tmp;
+
+	if (bank == 0) {
+		tmp = *(gpio_common.base + dirm0);
+		*(gpio_common.base + dirm0) = (tmp & mask) | (dir & ~mask);
+	}
+	else if (bank == 1) {
+		tmp = *(gpio_common.base + dirm1);
+		*(gpio_common.base + dirm1) = (tmp & mask) | (dir & ~mask);
+	}
+	else {
+		return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int gpio_readPin(unsigned int bank, unsigned int pin, uint32_t *val)
+{
+	int err;
+
+	err = gpio_checkPin(bank, pin);
+	if (err < 0) {
+		return err;
+	}
+
+	err = gpio_readPort(bank, val);
+	if (err < 0) {
+		return err;
+	}
+
+	*val = !!(*val & (1 << pin));
+
+	return EOK;
+}
+
+
+int gpio_writePin(unsigned int bank, unsigned int pin, uint32_t val)
+{
+	unsigned char word = 0;
+	int err;
+
+	err = gpio_checkPin(bank, pin);
+	if (err < 0) {
+		return err;
+	}
+
+	if (pin >= 16) {
+		pin -= 16;
+		word = 1;
+	}
+
+	return gpio_writePort(bank, word, ((~(1 << pin)) << 16) | ((!!val) << pin));
```

Comment:
MISRA: sadly `!!val` is a not allowed cast between different essential types (int -> bool -> int)

Reply:
Fixed, now I'm using `(val) ? 1 : 0`.